File: leetcode/0001__two_sum.py
"""
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.
"""
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        # return self.twoSum1(nums, target)
        # return self.twoSum2(nums, target)
        return self.twoSumWithDict(nums, target)

    def twoSum1(self, nums: list[int], target: int) -> list[int]:
        """
        Naive solution: Starting from first number, try all the other numbers
        following it, until a solution is found. If none found, go to next number.

        Time: O(n^2) - test n elements at most n times
        Space: O(1) - no additional space needed

        >>> s = Solution()
        >>> sorted(s.twoSum1([2, 7, 11, 15], 9))
        [0, 1]
        >>> sorted(s.twoSum1([3, 2, 4], 6))
        [1, 2]
        >>> sorted(s.twoSum1([3, 3], 6))
        [0, 1]
        """
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i, j]
        return None

    def twoSum2(self, nums: list[int], target: int) -> list[int]:
        """
        1. Create a list of tuples from num: [num, index_of_num]
        2. Sort list
        3. For each number n, use binary search to see if there is another
           number that when added to n will give target; if not, continue

        Time:
        - List of tuples: O(n)
        - Sort: O(n*lg(n))
        - Binary search: O(lg(n)), at most n times
            - => O(n) + O(n*lg(n)) + O(n*lg(n))
            - => O(n*lg(n)) (for sufficiently large n, O(n) goes away)

        Space: O(n) for list of tuples

        >>> s = Solution()
        >>> sorted(s.twoSum2([2, 7, 11, 15], 9))
        [0, 1]
        >>> sorted(s.twoSum2([3, 2, 4], 6))
        [1, 2]
        >>> sorted(s.twoSum2([3, 3], 6))
        [0, 1]
        """
        tuples = []
        for i in range(len(nums)):
            tuples.append([nums[i], i])

        tuples.sort(key = lambda elt: elt[0])

        for elt in tuples:
            # binary search to find elt
            foundElt = self.binarySearch(tuples, [target - elt[0], elt[1]])
            if foundElt is not None:
                return [elt[1], foundElt[1]]
        
        return None

    def binarySearch(self, tuples: list[tuple[int, int]], targetTuple: tuple[int, int]) -> tuple[int, int]:
        """
        Binary search on tuples. Assumes tuples is sorted on the first element.
        """
        bottom = 0
        top = len(tuples) - 1
        mid = (top - bottom) // 2
        while bottom <= top:
            mid = (top + bottom) // 2
            if tuples[mid][0] == targetTuple[0]:
                break
            elif tuples[mid][0] > targetTuple[0]:
                top = mid - 1
            else:
                bottom = mid + 1

        if tuples[mid][0] == targetTuple[0]:
            # test indexes
            if tuples[mid][1] != targetTuple[1]:
                return tuples[mid]
            elif mid > 0 and tuples[mid - 1][0] == targetTuple[0]:
                return tuples[mid - 1]
            elif mid < len(tuples) - 2 and tuples[mid + 1][0] == targetTuple[0]:
                return tuples[mid + 1]
        else:
            return None

    def twoSumWithDict(self, nums: list[int], target: int) -> list[int]:
        """
        Create a dictionary mapping from value to its corresponding
        index in nums. For each value in nums, look up in dictionary
        the balance (value that is the difference of target and the
        ith value in nums). If it exists, return the pair of indices.

        Time:
        - let n = len(nums)
        - O(n) to generate dict
        - O(n) to iterate through nums, and for each iteration O(1) 
          dict lookup to check for its counterpart
        - => O(n)

        Space:
        - let n = len(nums)
        - O(n) dict mapping value to its index
        => O(n)
        """
        m = dict()
        for i in range(len(nums)):
            m[nums[i]] = i
        
        for i in range(len(nums)):
            balance = target - nums[i]
            if balance in m and m[balance] != i:
                return [i, m[balance]]
        return [-1, -1]

if __name__ == '__main__':
    import doctest
    doctest.testmod()

File: leetcode-ruby/0001__two_sum.rb
# 1. Two Sum
#
# Given an array of integers nums and an integer target, return indices of
# the two numbers such that they add up to target.
#
# You may assume that each input would have exactly one solution, and you
# may not use the same element twice.
#
# You can return the answer in any order.
#
# Example 1:
#
# Input: nums = [2,7,11,15], target = 9
# Output: [0,1]
# Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
#
# Example 2:
# 
# Input: nums = [3,2,4], target = 6
# Output: [1,2]
#
# Example 3:
#
# Input: nums = [3,3], target = 6
# Output: [0,1]

# Constraints:

# - 2 <= nums.length <= 10^4
# - -10^9 <= nums[i] <= 10^9
# - -10^9 <= target <= 10^9
# - Only one valid answer exists.

# Follow-up: Can you come up with an algorithm that is less than O(n^2)
# time complexity?

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
  return two_sum_2(nums, target)
end

# Naive two-loop version.
#
# Time: O(n^2)
# Space: O(1)
#
def two_sum_1(nums, target)
  i = 0
  while i < nums.length do
    j = i + 1
    while j < nums.length do
      if nums[i] + nums[j] == target
        return [i, j]
      end
      j += 1
    end
    i += 1
  end
end

# Same as above, but using Ranges instead of incrementing indexs.
#
def two_sum_1_1(nums, target)
  for i in (0 .. nums.length - 1).to_a
    for j in ((i + 1) .. (nums.length - 1)).to_a
      if nums[i] + nums[j] == target
        return [i, j]
      end
    end
  end
end

# Map of nums[i] => i for all i in nums.
#
# Time: O(n) - at most once through nums
# Space: O(n) - at most storing a value for each num in nums
#
def two_sum_2(nums, target)
  diffs = {}
  for i in (0 .. nums.length - 1).to_a
    found = diffs[nums[i]]
    if found != nil
      return [i, found]
    end
    diffs[target - nums[i]] = i
  end
end
